File: 9. Stacks/AsteroidCollision.js
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function(asteroids) {
    let stack = [];

    for (let asteroid of asteroids) {
        let destroyed = false;

        while (
            stack.length > 0 &&
            asteroid < 0 &&
            stack[stack.length - 1] > 0
        ) {
            let top = stack[stack.length - 1];

            if (top < -asteroid) {
                stack.pop(); // top asteroid explodes
                continue;
            } else if (top === -asteroid) {
                stack.pop(); // both explode
                destroyed = true;
                break;
            } else {
                // current asteroid is destroyed
                destroyed = true;
                break;
            }
        }

        if (!destroyed) {
            stack.push(asteroid);
        }
    }

    return stack;
};

File: 9. Stacks/AsteroidCollision.java
import java.util.*;

class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> st = new Stack<>();

        for (int asteroid : asteroids) {
            boolean destroyed = false;

            while (!st.isEmpty() && asteroid < 0 && st.peek() > 0) {
                if (st.peek() < -asteroid) {
                    st.pop(); // top asteroid explodes
                    continue; // keep checking the next top
                } else if (st.peek() == -asteroid) {
                    st.pop(); // both explode
                    destroyed = true;
                    break;
                } else {
                    // incoming asteroid is destroyed
                    destroyed = true;
                    break;
                }
            }

            if (!destroyed) {
                st.push(asteroid);
            }
        }

        // Convert stack to array
        int[] result = new int[st.size()];
        for (int i = st.size() - 1; i >= 0; i--) {
            result[i] = st.pop();
        }
        return result;
    }
}
